# NLP HW 1

### Niv Aharon Cohen and Omri 

#### 1+2.

In [115]:
#import torch
import nltk
import spacy
from pathlib import Path
import requests
from bs4 import BeautifulSoup
import pandas as pd

# Our libraries
import utils
import text_functions
from  text_functions import nltk_tokenize, spacy_tokenize, nltk_lemmatize, spacy_lemmatize, nltk_stemming


#### 3.

In [105]:
# LOAD DATA
spam_df = utils.load_dataset("spam.csv")

spam_df.head()

label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

#### 4.

In [106]:
# Calculate data statistics
stats = utils.calculate_statistics(spam_df)
utils.display_statistics(stats)

Total SMS messages: 5572
Number of spam messages: 747
Total word count: 90106
Average number of words per message: 16.17
5 most frequent words: [('i', 3001), ('to', 2242), ('you', 2240), ('a', 1433), ('the', 1328)]
Number of rare words: 4376


#### 5.

In [107]:
#### Loading necessary packages ####


nltk_data_path = Path(r"C:\\Users\\97254\\Desktop\\niv\\AFEKA\\M.sc Machine learning\\AI 2th year\\NLP\\NLP HOMEWORK\\NLP_HW1")  # Adjust this path if necessary

# Ensure the path exists
if not nltk_data_path.exists():
    print(f"Warning: The path {nltk_data_path} does not exist!")

# Add the path to NLTK's data path
nltk.data.path.append(str(nltk_data_path))

# Ensure 'punkt' is downloaded to the specified location
nltk.download('punkt', download_dir=str(nltk_data_path))
nltk.download('punkt_tab',download_dir=str(nltk_data_path))


# Load the spaCy language model
nlp = spacy.load("en_core_web_sm")

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\97254\Desktop\niv\AFEKA\M.sc Machine
[nltk_data]     learning\AI 2th year\NLP\NLP HOMEWORK\NLP_HW1...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\97254\Desktop\niv\AFEKA\M.sc Machine
[nltk_data]     learning\AI 2th year\NLP\NLP HOMEWORK\NLP_HW1...
[nltk_data]   Package punkt_tab is already up-to-date!


In [113]:

text_line = spam_df['message'].iloc[10]  # Modify the index if needed
print("Original text:")
print(text_line, "\n")

# Tokenize using NLTK
nltk_tokens = nltk_tokenize(text_line)
print("NLTK Tokenization:")
print(nltk_tokens)

# Tokenize using spaCy
spacy_tokens = spacy_tokenize(text_line, nlp)
print("\nspaCy Tokenization:")
print(spacy_tokens)


Original text:
I'm gonna be home soon and i don't want to talk about this stuff anymore tonight, k? I've cried enough today. 

NLTK Tokenization:
['I', "'m", 'gon', 'na', 'be', 'home', 'soon', 'and', 'i', 'do', "n't", 'want', 'to', 'talk', 'about', 'this', 'stuff', 'anymore', 'tonight', ',', 'k', '?', 'I', "'ve", 'cried', 'enough', 'today', '.']

spaCy Tokenization:
['I', "'m", 'gon', 'na', 'be', 'home', 'soon', 'and', 'i', 'do', "n't", 'want', 'to', 'talk', 'about', 'this', 'stuff', 'anymore', 'tonight', ',', 'k', '?', 'I', "'ve", 'cried', 'enough', 'today', '.']


In [109]:
# Tokenize the entire dataframe
nltk_tokens_set = text_functions.create_token_set_nltk(spam_df)
spacy_tokens_set = text_functions.create_token_set_spacy(spam_df, nlp)
# Output the results
print("Number of unique tokens (NLTK):", len(nltk_tokens_set))
print("Number of unique tokens (spaCy):", len(spacy_tokens_set))

print("Number of unique tokens (NLTK):", nltk_tokens_set)
print("Number of unique tokens (spaCy):", spacy_tokens_set)

Number of unique tokens (NLTK): 11520
Number of unique tokens (spaCy): 11582
Number of unique tokens (NLTK): {'So', '+447797706009', '8552', 'middle', 'Euro2004', 'coccooning', '07801543489', '.........', 'Always', 'deposited', 'love.rakhesh', 'Welcome', 'Like', 'near', 'Mobileupd8', 'yer', 'To', 'quoting', 'X49', 'fit', 'WOT', 'cc100p/min', 'shopping', 'CUM', 'lim', 'LONGER', 'remembered', 'F', 'tagged', 'å£1450', 'FRAN', 'citizen', 'Unni', 'silently', 'umma', 'sips', 'parchi', 'WAY', 'MUCHXXLOVE', '7634', 'Jamster', 'Forgets', 'FEEL', 'token', 'Congrats', 'recount', 'tv/', 'spring', 'Latest', 'u.Othrwise', 'ore', 'landline', "I'ma", 'eerie', 'this', 'k.', 'age', 'keys', 'fool', 'armenia', '62735=å£450', '08719181259', 'edhae', 'indian', 'Ummmmmaah', 'fault', 'asus', 'brilliant', 'celeb', 'strips', 'secure', 'interview', 'ASUSUAL', 'others', 'HOWS', 'count', '\\STOP', "'taxless", 'MAKE', 'Btw', 'ortxt', 'askd', '07742676969', 'Deliver', 'Shiny', 'water', 'yowifes', 'Paris', 'Wan', '08

#### Results: 
In the example from row 10 of the dataset , the tokenization output of NLTK and spaCy is similar overall. Although, we can see that by tokenizing the entire dataset, there is a difference in the size of the sets of NLTK and spaCy. Apparently, spaCy did additional tokenization to some of the words in the messages, thus, creating a bigger set than NLTK.

#### 6.

In [7]:
### Loading additional libreries from nltk ###

# Ensure NLTK data path and download WordNet data

nltk.data.path.append(str(nltk_data_path))
nltk.download('wordnet', download_dir=str(nltk_data_path))
nltk.download('omw-1.4', download_dir=str(nltk_data_path))  # Required for WordNet

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\97254\Desktop\niv\AFEKA\M.sc Machine
[nltk_data]     learning\AI 2th year\NLP\NLP HOMEWORK\NLP_HW1...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\97254\Desktop\niv\AFEKA\M.sc Machine
[nltk_data]     learning\AI 2th year\NLP\NLP HOMEWORK\NLP_HW1...
[nltk_data]   Package omw-1.4 is already up-to-date!


True

In [8]:
text_line = spam_df['message'].iloc[10]  # Modify the index if needed
print("Original Text:")
print(text_line)

# Perform NLTK lemmatization
nltk_lemmas = nltk_lemmatize(text_line)
print("\nNLTK Lemmatization:")
print(nltk_lemmas)

# Perform spaCy lemmatization
spacy_lemmas = spacy_lemmatize(text_line, nlp)
print("\nspaCy Lemmatization:")
print(spacy_lemmas)

Original Text:
I'm gonna be home soon and i don't want to talk about this stuff anymore tonight, k? I've cried enough today.

NLTK Lemmatization:
['I', "'m", 'gon', 'na', 'be', 'home', 'soon', 'and', 'i', 'do', "n't", 'want', 'to', 'talk', 'about', 'this', 'stuff', 'anymore', 'tonight', ',', 'k', '?', 'I', "'ve", 'cried', 'enough', 'today', '.']

spaCy Lemmatization:
['I', 'be', 'go', 'to', 'be', 'home', 'soon', 'and', 'I', 'do', 'not', 'want', 'to', 'talk', 'about', 'this', 'stuff', 'anymore', 'tonight', ',', 'k', '?', 'I', 'have', 'cry', 'enough', 'today', '.']


In [110]:
nltk_lemmas_set, spacy_lemmas_set = text_functions.create_lemma_set(nltk_tokens_set, spacy_tokens_set, nlp)
# Output the results
print("Number of unique lemmas (NLTK):", len(nltk_lemmas_set))
print("Number of unique lemmas (spaCy):", len(spacy_lemmas_set))

print(nltk_lemmas_set)
print(spacy_lemmas_set)


Number of unique lemmas (NLTK): 11052
Number of unique lemmas (spaCy): 9843
{'So', '+447797706009', 'middle', '8552', 'Euro2004', 'coccooning', '07801543489', '.........', 'Always', 'deposited', 'love.rakhesh', 'Welcome', 'invoice', 'Like', 'near', 'Mobileupd8', 'yer', 'To', 'quoting', 'X49', 'fit', 'WOT', 'cc100p/min', 'implication', 'shopping', 'CUM', 'lim', 'LONGER', 'remembered', 'F', 'tagged', 'å£1450', 'FRAN', 'citizen', 'Unni', 'silently', 'umma', 'parchi', 'WAY', 'MUCHXXLOVE', '7634', 'Jamster', 'Forgets', 'FEEL', 'token', 'Congrats', 'recount', 'tv/', 'spring', 'Latest', 'u.Othrwise', 'ore', 'landline', "I'ma", 'eerie', 'this', 'k.', 'age', 'armenia', 'fool', '62735=å£450', '08719181259', 'suck', 'edhae', 'indian', 'Ummmmmaah', 'fault', 'asus', 'brilliant', 'celeb', 'secure', 'interview', 'ASUSUAL', 'others', 'HOWS', 'count', '\\STOP', "'taxless", 'MAKE', 'Btw', 'ortxt', 'askd', '07742676969', 'Deliver', 'Shiny', 'water', 'yowifes', 'Paris', 'Wan', '087104711148', 'boughtåÓbra

#### Results:

Difference between the two libraries is: 

NLTK results in tokens that are closer to the original text and keeps contractions split into components, for example, "'m" stayes the same. 

Spacy resolves contractions into their expanded or semantic forms and reduces words to their base forms, for example, "'m" becomes "be", "na" becomes "to".

We see that NLTK's set is larger than spaCy's set, this is because NLTK's lemmatizer (WordNetLemmatizer) relies on the WordNet lexical database, which sometimes returns the input token itself when no clear lemma is found. This behavior can result in more lemmas.
spaCy's lemmatizer has its own robust set of rules and models, which might consolidate more tokens into the same lemma, reducing the total count of unique lemmas.

We can also observe that the length of the set decreased after the lemmatization process. This is because lemmatization reduces words to their base form, which can lead to duplicates from different forms of the same word. These duplicates are discarded, resulting in a smaller set.

#### 7.

(spaCy library does not have stemming method)


In [10]:
text_line = spam_df['message'].iloc[10]  # Modify the index if needed
nltk_stemmed = nltk_stemming(text_line)


print("Original Text:", text_line)
print("\nNLTK Stemming:", nltk_stemmed)


Original Text: I'm gonna be home soon and i don't want to talk about this stuff anymore tonight, k? I've cried enough today.

NLTK Stemming: ["i'm", 'gonna', 'be', 'home', 'soon', 'and', 'i', "don't", 'want', 'to', 'talk', 'about', 'thi', 'stuff', 'anymor', 'tonight,', 'k?', "i'v", 'cri', 'enough', 'today.']


In [112]:
nltk_stemmed_set = text_functions.nltk_set_stemming(nltk_tokens_set)
print("Number of unique stemms (NLTK):", len(nltk_stemmed_set))

print(nltk_stemmed_set)

Number of unique stemms (NLTK): 8110
{'everyon', '+447797706009', '8552', '07801543489', '1yf', '.........', 'love.rakhesh', 'near', 'yer', 'reappli', 'cheesi', 'fit', 'cc100p/min', 'lim', 'å£1450', 'announc', 'citizen', 'everyth', 'umma', 'parchi', 'alrit', 'hurri', 'ahhh', '7634', 'tire', 'ref:9280114', 'shape', 'unintent', 'token', 'cheeto', 'recount', 'reppurcuss', 'prap', 'tv/', 'spring', 'ore', 'noooooooo', 'k.', 'depart', 'age', 'condit', 'armenia', 'fool', '62735=å£450', 'astoundingli', '08719181259', 'hm', 'improv', 'suck', 'indian', 'australia', 'fault', 'insid', 'brilliant', 't.', 'celeb', 'dena', 'whatsup', 'interview', 'count', "'taxless", 'ortxt', 'askd', '07742676969', 'water', 'me.ne', 'spook', '087104711148', 'onward', 'chennai', 'anna', 'senor', 'tb', 'textbook', 'bullshit', 'physic', 'xxx\\', 'ringtoneking.co.uk', '050703', 'eldest', 'rich', 'that2worzel', 'project', 'box245c2150pm', 'rumour', 'hogli', '13/10/04', 'ride', 'varaya', 'yourself', 'prasad', '5.00', '4.47

#### Results:

 NLTK stemming truncates words to their root forms without regard for their linguistic correctness, often producing incomplete or non-dictionary words, such as "thi" (from "this") and "cri" (from "cried").  
 
 The NLTK stemming process reduced the number of unique tokens to 8110. This reduction occurs because stemming removes affixes from words, often truncating them to a common root form. For example, "running", "runner", and "ran" may all be reduced to "run". As a result, different word forms that belong to the same root or word family are collapsed into a single stem, reducing the overall number of unique tokens.

#### 8.

In [12]:
message_to_remove = text_functions.find_message_to_remove_stemm_less_lemma_equal(spam_df)

if message_to_remove:
    print("Message to remove:", message_to_remove)


No message found that satisfies the condition.


##### Results: 

Task 8 requires a message where stemming removes tokens, but lemmatization does not. This is unlikely because stemming reduces tokens more aggressively than lemmatization.

#### 9.

In [13]:
message_to_remove_9 = text_functions.find_message_to_remove_lemma_less_stemm_equal(spam_df)

if message_to_remove_9:
    print("Message to remove:", message_to_remove_9)

No message found that satisfies the condition.


##### Results:

Task 9 requires a message where lemmatization reduces tokens, but stemming does not. This is also unlikely because lemmatization is more conservative, and any reduction in lemmatized tokens would likely also reduce the count of stemmed tokens.

#### 10.

NLTK's tokenization breaks the text into individual components, such as words, punctuation marks, and contractions, while preserving their original forms. For example, "I'm" becomes ["I", "'m"], "gonna" becomes ["gon", "na"], and punctuation like "?" and "." are treated as separate tokens.

Lemmatization refines the tokens into their dictionary forms (lemmas) based on grammatical context. In this case, lemmatization yields the same result as tokenization because the words in the text are already in their base forms (e.g., "cried" remains "cried" as it's a past-tense verb in its base inflection).

Stemming, on the other hand, reduces words to their root forms by truncating suffixes or prefixes. This approach is less context-sensitive and often creates non-standard outputs. For instance, "cried" is stemmed to "cri," "anymore" to "anymor," and "this" to "thi." Unlike lemmatization, stemming disregards grammatical correctness and focuses solely on stripping affixes, sometimes leading to distorted forms.

Overall, tokenization preserves the structure, lemmatization maintains grammatical accuracy, and stemming sacrifices precision for simplicity, making it less suitable for nuanced text processing.

#### 11.

##### Tokenization:
SpaCy's tokenization splits the text into individual components, similar to NLTK. It identifies contractions like "I'm" as separate tokens (['I', "'m"]) and splits colloquial expressions like "gonna" into ['gon', 'na']. Punctuation marks are also treated as standalone tokens, such as ',', '?', and '.'. This tokenization aligns closely with NLTK, as it retains the text's structure without modifications.

##### Lemmatization:
SpaCy's lemmatization refines tokens into their base forms or lemmas, taking into account grammatical context. This approach is more sophisticated than NLTK's, as seen in several examples:

"gon" (from "gonna") is lemmatized to "go," capturing its intended meaning.
"n't" is transformed into "not," ensuring semantic clarity.
"cried" becomes "cry," its base verb form.
"I've" splits into "I" and "have," preserving grammatical accuracy.
Unlike NLTK, SpaCy applies grammatical rules to produce more meaningful base forms, such as converting contractions and colloquialisms into their standard equivalents.

##### Summary:
SpaCy excels in lemmatization by providing grammatically accurate and meaningful base forms, surpassing NLTK in understanding the context. Both libraries perform similarly in tokenization, splitting text into logical units while retaining the original structure. For stemming, NLTK offers a basic tool, but SpaCy's advanced lemmatization effectively makes stemming unnecessary in many use cases.

#### 13-15.

In [ ]:
import re

url = 'https://edition.cnn.com/2024/12/06/business/google-pay-lawsuit-cfpb/index.html'

# Send a GET request to the page
response = requests.get(url)

# Parse the HTML content
soup = BeautifulSoup(response.content, 'html.parser')

# Remove unwanted sections (footers, ads, etc.)
for unwanted_section in soup(['footer', 'aside', 'div', 'span', 'figcaption', 'figure']):
    # You can refine this further by filtering specific classes or ids
    if 'advertisement' in unwanted_section.get('class', []):
        unwanted_section.decompose()  # Remove the section

# Find the main article content (you may need to inspect the HTML structure to find the right tag)
article_content = soup.find('div', {'class': 'article-body'})  # Adjust as per the website structure

# If you cannot find the main article content directly, look for alternative tags/classes like 'content', 'main', etc.
if not article_content:
    article_content = soup.find('main')

# Extract and clean the article text
if article_content:
    clean_text = article_content.get_text()
    #print(clean_text)  # Print or process the article content
else:
    print("Article content not found")


# Find the index of the word "CNN" and extract text from there
start_index = clean_text.find("CNN")
if start_index != -1:
    extracted_text = clean_text[start_index:]
else:
    extracted_text = "The word 'CNN' was not found in the text."

# Remove excessive new lines from text
extracted_text = extracted_text.replace("\n", "")

# Remove newlines and normalize spacing
cleaned_text = " ".join(extracted_text.split())

# Print the extracted text
print(cleaned_text)


CNN — Google sued the Consumer Financial Protection Bureau on Friday, challenging the agency’s decision to place Google’s payment division under federal supervision. In a copy of the lawsuit provided by Google, the company said the CFPB’s supervision would be a “burdensome form of regulation” imposed based on a “small number of unsubstantiated user complaints.” The CFPB’s decision related to a Google peer-to-peer payment product no longer offered in the United States. The lawsuit, filed in the US district court in Washington, DC, comes after the CFPB published an order announcing supervisory authority of Google Payment Corp. The agency alleged that Google’s handling of its payment products may pose a risk to consumers. The CFPB cited customer complaints, including that Google failed to properly investigate instances where money was transferred in error. The legal fight between Google and the CFPB, the government agency founded to enforce consumer protection laws, comes amid a push by b

In [154]:
# Split the text into sentences by "."
sentences_url = [sentence.strip() for sentence in cleaned_text.split('.') if sentence.strip()]

# Create a DataFrame
df_clean_text = pd.DataFrame(sentences_url, columns=["message"])

# Display the DataFrame
print(df_clean_text)

                                              message
0   CNN — Google sued the Consumer Financial Prote...
1   In a copy of the lawsuit provided by Google, t...
2   ” The CFPB’s decision related to a Google peer...
3   The lawsuit, filed in the US district court in...
4   The agency alleged that Google’s handling of i...
5   The CFPB cited customer complaints, including ...
6   The legal fight between Google and the CFPB, t...
7   In a statement, Google spokesperson José Casta...
8   “This is a clear case of government overreach ...
9                                                   S
10     , and we are challenging it in court,” he said
11  In the company’s lawsuit, Google argued that t...
12  “As a matter of common sense, a product that n...
13  However, the CFPB said the discontinuation of ...
14  Supervisory authority over Google’s payments d...
15  In Google’s lawsuit, the company alleged the C...
16  In 2022, the CFPB announced it would begin exa...
17  “This authority gives us

#### 15.5

In [155]:

# Tokenize the entire dataframe
nltk_tokens_set_url = text_functions.create_token_set_nltk(df_clean_text)
spacy_tokens_set_url = text_functions.create_token_set_spacy(df_clean_text, nlp)
# Output the results
print("Number of unique url tokens (NLTK):", len(nltk_tokens_set_url))
print("Number of unique url tokens (spaCy):", len(spacy_tokens_set_url))


print(nltk_tokens_set_url)
print(spacy_tokens_set_url)

Number of unique url tokens (NLTK): 214
Number of unique url tokens (spaCy): 214
{'was', 'are', 'the', 'Friday', 'information', 'Chopra', 'small', '2022', 'tech', 'agency', 'As', 'companies', 'exists', 'incapable', 'order', 'Corp', 'market', 'unsubstantiated', 'and', 'matter', 'such', 'comply', 'us', 'that', 'details', 'payment', 'contributed', 'US', 'not', 'However', 'allowing', 'risk', 'they', 'posing', 'bar', 'nonbank', 'operations', 'discontinuation', 'of', 'requests', 'consumer', 'spokesperson', 'subject', '“', ',', 'company', 'sense', 'context', 'report', 'been', 'in', '’', 'updated', 'with', 'between', 'after', 'Google', 'form', 'offered', 'Clare', 'Duffy', 'CFPB', 'a', 'cited', 'protection', 'peer-to-peer', 'imposed', 'move', 'we', 'case', 'supervisory', 'on', 'Supervisory', 'enforce', 'sued', 'decision', 'examining', 'number', 'regulation', 'lawsuit', 'States', 'division', 'money', 'Castañeda', 'U', 'from', 'In', 'Washington', 'federal', 'investigate', 'setting', 'did', 'sprea

##### Results:
NLTK and spaCy produce nearly identical sets of tokens, with NLTK containing one additional token compared to spaCy.

#### 15.6


In [156]:
nltk_lemmas_set_url, spacy_lemmas_set_url = text_functions.create_lemma_set(nltk_tokens_set_url, spacy_tokens_set_url, nlp)
# Output the results
print("Number of unique lemmas (NLTK):", len(nltk_lemmas_set_url))
print("Number of unique lemmas (spaCy):", len(spacy_lemmas_set_url))

print(nltk_lemmas_set_url)
print(spacy_lemmas_set_url)

Number of unique lemmas (NLTK): 206
Number of unique lemmas (spaCy): 192
{'are', 'information', 'the', 'Friday', 'Chopra', 'small', '2022', 'examination', 'tech', 'agency', 'As', 'exists', 'incapable', 'wa', 'order', 'Corp', 'market', 'instance', 'institution', 'unsubstantiated', 'and', 'matter', 'such', 'comply', 'that', 'payment', 'contributed', 'US', 'not', 'However', 'allowing', 'risk', 'they', 'posing', 'bar', 'nonbank', 'discontinuation', 'of', 'consumer', 'spokesperson', 'subject', '“', ',', 'company', 'sense', 'document', 'context', 'been', 'updated', 'in', '’', 'with', 'between', 'after', 'Google', 'form', 'ha', 'offered', 'Clare', 'Duffy', 'CFPB', 'a', 'protection', 'peer-to-peer', 'imposed', 'move', 'we', 'case', 'supervisory', 'on', 'Supervisory', 'enforce', 'sued', 'decision', 'examining', 'number', 'regulation', 'lawsuit', 'States', 'division', 'money', 'Castañeda', 'U', 'from', 'In', 'Washington', 'federal', 'investigate', 'setting', 'did', 'story', 'count', 'to', 'posed

##### Results:

The comparison between NLTK and spaCy on the lemmas of the provided text reveals several differences. NLTK identified 207 unique lemmas, while spaCy identified 192, indicating that NLTK retains more variations as separate lemmas. For example, NLTK includes lemmas like “are” and “been” that are missing in spaCy, whereas spaCy includes lemmas like “base” and “handle” not present in NLTK. SpaCy appears to apply broader normalization, reducing words like “argued” to “argue” and “handled” to “handle,” while NLTK keeps them distinct. Additionally, punctuation is handled differently; NLTK retains symbols like ““” and “,” separately, while spaCy consolidates some punctuation and includes symbols like “—” as lemmas. Both tools handle proper nouns and acronyms, such as “CFPB” and “Google,” similarly, preserving them as is. Overall, NLTK preserves more granular distinctions, while spaCy emphasizes broader reductions for normalization.

#### 15.7


In [157]:
nltk_stemmed_set_url = text_functions.nltk_set_stemming(nltk_tokens_set_url)
print("Number of unique stemms (NLTK):", len(nltk_stemmed_set_url))

print(nltk_stemmed_set_url)

Number of unique stemms (NLTK): 191
{'ensur', 'are', 'base', 'the', 'clare', 'small', '2022', 'tech', 'divis', 'publish', 'provid', 'unsubstanti', 'overse', 'cfpb', 'incap', 'wa', 'order', 'market', 'rohit', 'file', 'argu', 'and', 'matter', 'such', 'us', 'that', 'announc', 'payment', 'howev', 'not', 'corp', 'risk', 'they', 'bar', 'addit', 'instanc', 'nonbank', 'dc', 'institut', 'of', 'josé', 'subject', 'spokesperson', '“', ',', 'document', 'involv', 'context', 'enforc', 'feder', 'been', 'in', '’', 'challeng', 'with', 'releas', 'between', 'after', 'form', 'supervis', 'ha', 'confidenti', 'suffici', 'a', 'on-sit', 'move', 'we', 'case', 'director', 'includ', 'on', 'relat', 'contribut', 'number', 'regul', 'lawsuit', 'consum', 'money', 'cite', 'inform', 'from', 'washington', 'agil', 'agenc', 'did', 'count', 'to', 'copi', 'critic', 'set', 'financi', 'investig', 'which', 'what', 'into', 'for', 'discontinu', 'peer-to-p', 'is', 'complaint', 'spread', 'samsung', 'amid', 'unit', 'low', 'fight', 'f

#### Results:
Stemming reduces the size of the set compared to lemmatization by transforming words to their root forms, effectively merging words with similar roots. This process often removes grammatical nuances, resulting in terms like "ensur" that lack complete grammatical meaning.

#### 15.8-9

In [158]:
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.tokenize import word_tokenize

# Function definition 
def find_message_to_remove(df):
    # Initialize stemmer and lemmatizer
    stemmer = PorterStemmer()
    lemmatizer = WordNetLemmatizer()


    # Function to preprocess and calculate unique tokens
    def get_unique_tokens(messages):
        stemmed_tokens = set()
        lemmatized_tokens = set()
        for msg in messages:
            tokens = word_tokenize(msg.lower())
            stemmed_tokens.update(stemmer.stem(token) for token in tokens)
            lemmatized_tokens.update(lemmatizer.lemmatize(token) for token in tokens)
        return stemmed_tokens, lemmatized_tokens

    # Calculate original token sets
    original_stemmed, original_lemmatized = get_unique_tokens(df["message"])

    #init flags and string holders for the message removed that make the condition true
    msg_less_stem_equal_lemma = ""
    msg_less_lemma_equal_stem = ""

    found_msg_1_flag = False
    found_msg_2_flag = False


    # Find the message to remove
    for idx, row in df.iterrows():
        # Remove one message
        remaining_messages = df.drop(index=idx)["message"]
        stemmed, lemmatized = get_unique_tokens(remaining_messages)
        
        # Check the conditions
        if len(stemmed) < len(original_stemmed) and len(lemmatized) == len(original_lemmatized):
            print(f"Message  at index {idx}  causes a change in stemmed tokens but not in lemmatized tokens. \nmessage: ", row["message"])
            #return row["message"]
            msg_less_stem_equal_lemma = row["message"] 
            found_msg_1_flag = True
        
        if len(lemmatized) < len(original_lemmatized) and len(stemmed) == len(original_stemmed):
             print(f"Message at index {idx} causes a change in lemmatized tokens but not in stemmed tokens. \nmessage:", row["message"])
            # return row["message"]  # Return the message that causes the change
             msg_less_lemma_equal_stem = row["message"] 
             found_msg_2_flag = True

        if (found_msg_1_flag == True and found_msg_2_flag== True):
            break # Break the loop if such messages found

            

    if  (found_msg_1_flag == False and found_msg_2_flag == False):      
        print("No message found that meets the criteria.")
    
  
   
    #return None

In [160]:
find_message_to_remove(df_clean_text)

No message found that meets the criteria.


#### Results:


The absence of a message in the dataset that meets the specified conditions could be due to the inherent differences between stemming and lemmatization processes. Stemming is a more aggressive method that removes affixes to produce a root form, often discarding grammatical context. Lemmatization, on the other hand, generates linguistically valid base forms by considering the word's context and part of speech. This divergence makes it unlikely for a message to exist where removing it alters only one token count while leaving the other unchanged. Additionally, the lack of such messages may reflect the dataset's specific characteristics, where no individual message contains terms that exhibit these unique processing behaviors.

#### 15.10

The comparison between NLTK tokens, lemmas, and stems highlights differences in linguistic processing and simplification. The tokenization process resulted in 214 unique tokens, which include raw words split by spaces and punctuation, preserving grammatical meaning but often containing redundancies such as variations in forms (e.g., "was" and "wa"). Lemmatization reduced the count to 206 unique lemmas by normalizing words to their base dictionary forms while retaining grammatical context (e.g., "are" remains unchanged). Stemming further simplified the text, reducing the count to 191 unique stems by stripping words to their root forms without considering grammatical correctness, leading to more aggressive reductions and potential loss of meaning (e.g., "ensuring" becomes "ensur"). This progression illustrates how lemmatization provides a balance between grammatical integrity and redundancy, while stemming focuses on simplification at the expense of linguistic nuance.


#### 15.11

The comparison of spaCy's tokenization and lemmatization reveals notable differences in the level of text simplification achieved. SpaCy's tokenization resulted in 214 unique tokens, which include raw words segmented from the text, preserving punctuation and grammatical structure but often containing duplicates or inflected forms (e.g., "are" and "is"). Lemmatization reduced this count to 192 unique lemmas by normalizing words to their base forms, eliminating redundancies while maintaining grammatical meaning (e.g., "was" and "were" become "be"). Compared to the tokenization step, lemmatization in spaCy offers a streamlined representation of the text by focusing on semantic equivalence, making it particularly useful for tasks requiring a deeper understanding of word meanings and relationships. The smaller number of lemmas compared to tokens reflects a more compact and meaningful vocabulary derived from the original text.

#### 16 - 17.


In [ ]:
whatsapp_text_file_path = 'whatsapp_chat.txt'

# Read the file content into the variable
with open(whatsapp_text_file_path, 'r', encoding='utf-8') as file:
    whatsapp_text = file.read()

# Print the content (optional)
print(whatsapp_text)


[13.2.2024, 19:26:47] ניב אפקה: ‏ההודעות והשיחות מוצפנות מקצה לקצה. לאף אחד מחוץ לצ'אט הזה, גם לא ל-WhatsApp, אין אפשרות לקרוא אותן ולהאזין להן.
‏[13.2.2024, 19:26:47] Omri: CamScanner 2024-02-13 18.57.pdf • ‏11 דפים ‏מסמך הושמט
[13.2.2024, 19:26:47] Omri: פתרון שלי להסתברות
[13.2.2024, 19:26:47] Omri: אם מצאת טעות על הדרך תעדכן
[13.2.2024, 22:29:55] ניב אפקה: איזה מלאך אתה❤️❤️
[13.2.2024, 22:30:33] ניב אפקה: טנקיו! מחר לא אוכל לבוא לצערי אז לא אוכל לארח לך חברה
[13.2.2024, 22:33:38] Omri: לא תהיה אווירה סטודנטיאלית
[13.2.2024, 22:40:58] ניב אפקה: חחחח
[21.2.2024, 17:28:21] ניב אפקה: אח שלי, תסלח לי אבל גם הפעם לא אבוא לארח לך חברה, אני איתך מרחוק, ראיתי שגם ברק לא בא לארח לך חברה חחח
[21.2.2024, 18:08:57] Omri: בדד אלך גם תפילה אין לי
[10.5.2024, 17:41:22] Omri: איפה אתה תגיד לי
[11.5.2024, 21:54:00] ניב אפקה: יא מלאך, שבוע טוב שיהיה לך. אני הייתי בזום השבועיים האלה, גם גנבו לי את הקורקינט אז לא היה לי ממש איך להגיע, אבל אגיע בתקווה בשבוע הבא בשביל באווירה הסטודנטיאלית😉
[11.5.2024, 21

In [136]:


# Define regex for splitting based on sender names

names_pattern = r"(?=\[\d{1,2}\.\d{1,2}\.\d{4}, \d{2}:\d{2}:\d{2}\] (Omri|ניב אפקה))"

# Split the text into messages
messages = re.split(names_pattern, whatsapp_text)

# Process messages into structured format
structured_messages = []
for i in range(1, len(messages), 2):  # Skip every alternate index (matched name)
    structured_messages.append(messages[i - 1] + messages[i])

# Extract message details into a DataFrame
message_pattern = r"\[(.*?)\] (.*?): (.*)"
data = []
for msg in structured_messages:
    match = re.match(message_pattern, msg.strip())
    if match:
        date_time, sender, content = match.groups()
        data.append({"DateTime": date_time, "Sender": sender, "message": content})

# Convert to DataFrame
df_whatsapp_processed_text = pd.DataFrame(data)

# Display the DataFrame
print(df_whatsapp_processed_text)

               DateTime    Sender  \
0   13.2.2024, 19:26:47  ניב אפקה   
1   13.2.2024, 19:26:47      Omri   
2   13.2.2024, 19:26:47      Omri   
3   13.2.2024, 19:26:47      Omri   
4   13.2.2024, 22:29:55  ניב אפקה   
..                  ...       ...   
92  1.12.2024, 12:21:37      Omri   
93  1.12.2024, 12:21:48      Omri   
94  1.12.2024, 12:21:57      Omri   
95  1.12.2024, 15:45:05  ניב אפקה   
96  1.12.2024, 16:41:17  ניב אפקה   

                                              message  
0   ‏ההודעות והשיחות מוצפנות מקצה לקצה. לאף אחד מח...  
1   CamScanner 2024-02-13 18.57.pdf • ‏11 דפים ‏מס...  
2                                  פתרון שלי להסתברות  
3                          אם מצאת טעות על הדרך תעדכן  
4                                   איזה מלאך אתה❤️❤️  
..                                                ...  
92                      אני גם רק עכשיו חוזר לעניינים  
93                                3 דייטים לא בא ברגל  
94                             זהו אני עושה הפסקה ק

#### 17.5

In [141]:

# Tokenize the entire dataframe
nltk_tokens_set_whatsapp = text_functions.create_token_set_nltk(df_whatsapp_processed_text)
spacy_tokens_set_whatsapp = text_functions.create_token_set_spacy(df_whatsapp_processed_text, nlp)
# Output the results
print("Number of unique url tokens (NLTK):", len(nltk_tokens_set_whatsapp))
print("Number of unique url tokens (spaCy):", len(spacy_tokens_set_whatsapp))


print(nltk_tokens_set_whatsapp)
print(spacy_tokens_set_whatsapp)

Number of unique url tokens (NLTK): 387
Number of unique url tokens (spaCy): 394
{'כאב', 'בסופש', 'לא', 'מקולקל', 'גנבו', 'ה', 'אין', 'אותו', 'זהו', '?', 'שם', 'הקורקינט', 'יעבור', 'ברגל', 'הקו', 'משהו', 'תעשה', 'לקצה', '\u200fמסמך', 'בסוודר', 'להן', 'עוונות', 'לנו', ',', 'פרויקט', 'הייתי', 'שבוע', 'שבאמת', 'מצליח', 'היה', 'הכי', 'ולהאזין', 'להתחיל', 'מסיים', 'מחופשתי', 'שלחתי', 'ל', 'קומיט', 'ומחר', 'חחח', 'You', 'פוש', 'תגיד', 'מקצה', 'אשאל', 'תסלח', 'יותר', 'השבוע', 'גיט', 'פרוייקט', 'אחד', 'גם', 'פתרון', 'סיימו', "לצ'אט", 'התבלבלתי', 'נורמלי', 'מאפקה', 'מתכוון', 'כפרת', 'בן', 'LLM', 'האלה', 'פחות', 'מחוץ', 'אותן', 'בדד', 'אגיע', 'פורח', 'אהלן', 'הפרויקט', '\u200fסטיקר', 'אשלח', 'היו', 'צריכים', 'מרחוק', 'לוודא', 'ביחד', 'מרגש', 'החלומות', 'להתחבר', 'מאותו', 'חבר', 'ריענון', 'חברה', 'בלבולים', 'בשביל', 'מתוך', '\u200f7', 'הושמטה', 'בינתיים', '\u200f11', 'הקורקינט🙈', 'עם', 'הפרוייקטים', 'איך', 'שלומך', 'טוב', 'המקום', 'אז', 'תראה', 'הוטל', 'Attention', 'חינם', 'חזרתי', 'בית', 'זה', '

Differences of tokenization between NLTK and spaCy: 

NLTK often treats punctuation as standalone tokens and may have limited support for modern elements like emojis, while spaCy is better equipped for handling such characters, treating emojis and symbols as distinct tokens. spaCy also combines or interprets certain adjacent special characters and compound tokens more effectively than NLTK. These differences make spaCy preferable for nuanced, modern applications and NLTK better for traditional text processing.

#### 17.6

In [142]:
nltk_lemmas_set_whatsapp, spacy_lemmas_set_whatsapp = text_functions.create_lemma_set(nltk_tokens_set_whatsapp, spacy_tokens_set_whatsapp, nlp)
# Output the results
print("Number of unique lemmas (NLTK):", len(nltk_lemmas_set_whatsapp))
print("Number of unique lemmas (spaCy):", len(spacy_lemmas_set_whatsapp))

print(nltk_lemmas_set_whatsapp)
print(spacy_lemmas_set_whatsapp)


Number of unique lemmas (NLTK): 387
Number of unique lemmas (spaCy): 394
{'כאב', 'בסופש', 'לא', 'מקולקל', 'גנבו', 'ה', 'אין', 'אותו', 'זהו', '?', 'שם', 'הקורקינט', 'יעבור', 'ברגל', 'הקו', 'משהו', 'תעשה', 'לקצה', '\u200fמסמך', 'בסוודר', 'להן', 'עוונות', 'לנו', ',', 'פרויקט', 'הייתי', 'שבוע', 'מצליח', 'שבאמת', 'היה', 'הכי', 'ולהאזין', 'להתחיל', 'מסיים', 'מחופשתי', 'שלחתי', 'ל', 'קומיט', 'ומחר', 'חחח', 'You', 'פוש', 'תגיד', 'מקצה', 'אשאל', 'תסלח', 'יותר', 'השבוע', 'גיט', 'פרוייקט', 'אחד', 'גם', 'פתרון', 'סיימו', "לצ'אט", 'התבלבלתי', 'נורמלי', 'מאפקה', 'מתכוון', 'כפרת', 'בן', 'LLM', 'האלה', 'פחות', 'מחוץ', 'אותן', 'בדד', 'אגיע', 'פורח', 'אהלן', 'הפרויקט', '\u200fסטיקר', 'אשלח', 'היו', 'צריכים', 'מרחוק', 'לוודא', 'ביחד', 'מרגש', 'החלומות', 'להתחבר', 'מאותו', 'חבר', 'ריענון', 'חברה', 'בלבולים', 'בשביל', 'מתוך', '\u200f7', 'הושמטה', 'בינתיים', '\u200f11', 'הקורקינט🙈', 'עם', 'הפרוייקטים', 'איך', 'שלומך', 'טוב', 'המקום', 'אז', 'תראה', 'הוטל', 'Attention', 'חינם', 'חזרתי', 'בית', 'זה', 'לי', 'נו

#### 17.7

In [92]:
nltk_stemmed_set_whatsapp = text_functions.nltk_set_stemming(nltk_tokens_set_whatsapp)
print("Number of unique stemms (NLTK):", len(nltk_stemmed_set_whatsapp))

print(nltk_stemmed_set_whatsapp)

Number of unique stemms (NLTK): 387
{'כאב', 'בסופש', 'לא', 'מקולקל', 'גנבו', 'ה', 'אין', 'אותו', 'זהו', '?', 'שם', 'הקורקינט', 'יעבור', 'ברגל', 'הקו', 'משהו', 'תעשה', 'לקצה', '\u200fמסמך', 'בסוודר', 'להן', 'עוונות', 'לנו', 'llm', ',', 'פרויקט', 'הייתי', 'שבוע', 'מצליח', 'שבאמת', 'היה', 'הכי', 'ולהאזין', 'להתחיל', 'מסיים', 'ל-whatsapp', 'מחופשתי', 'שלחתי', 'ל', 'קומיט', 'ומחר', 'חחח', 'פוש', 'תגיד', 'מקצה', 'אשאל', 'תסלח', 'יותר', 'השבוע', 'גיט', 'פרוייקט', 'אחד', 'is', 'גם', 'פתרון', 'סיימו', "לצ'אט", 'התבלבלתי', 'נורמלי', 'מאפקה', 'מתכוון', 'כפרת', 'בן', 'האלה', 'פחות', 'מחוץ', 'אותן', 'בדד', 'אגיע', 'פורח', 'אהלן', 'הפרויקט', '\u200fסטיקר', 'אשלח', 'you', 'היו', 'צריכים', 'מרחוק', 'לוודא', 'ביחד', 'מרגש', 'החלומות', 'parsing.pi', 'להתחבר', 'מאותו', 'חבר', 'ריענון', 'חברה', 'בלבולים', 'בשביל', 'מתוך', '\u200f7', 'הושמטה', 'בינתיים', '\u200f11', 'הקורקינט🙈', 'עם', 'הפרוייקטים', 'איך', 'שלומך', 'טוב', 'המקום', 'אז', 'תראה', 'הוטל', 'חינם', 'חזרתי', 'בית', 'זה', 'לי', 'נושאים', 'בכוננות'

#### Results of 17.6 and 17.7:

Both NLTK and spaCy appear to lack support for lemmatization and stemming in the Hebrew language. As a result, their functionality for Hebrew text is limited to tokenization, producing identical outputs for tokenization, lemmatization, and stemming operations.

#### 17.8-9

In [102]:
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.tokenize import word_tokenize

# Function definition
def find_message_to_remove(df):
    # Initialize stemmer and lemmatizer
    stemmer = PorterStemmer()
    lemmatizer = WordNetLemmatizer()


    # Function to preprocess and calculate unique tokens
    def get_unique_tokens(messages):
        stemmed_tokens = set()
        lemmatized_tokens = set()
        for msg in messages:
            tokens = word_tokenize(msg.lower())
            stemmed_tokens.update(stemmer.stem(token) for token in tokens)
            lemmatized_tokens.update(lemmatizer.lemmatize(token) for token in tokens)
        return stemmed_tokens, lemmatized_tokens

    # Calculate original token sets
    original_stemmed, original_lemmatized = get_unique_tokens(df["message"])

    #init flags and string holders for the message removed that make the condition true
    msg_less_stem_equal_lemma = ""
    msg_less_lemma_equal_stem = ""

    found_msg_1_flag = False
    found_msg_2_flag = False


    # Find the message to remove
    for idx, row in df.iterrows():
        # Remove one message
        remaining_messages = df.drop(index=idx)["message"]
        stemmed, lemmatized = get_unique_tokens(remaining_messages)
        
        # Check the conditions
        if len(stemmed) < len(original_stemmed) and len(lemmatized) == len(original_lemmatized):
            print(f"Message  at index {idx}  causes a change in stemmed tokens but not in lemmatized tokens. \nmessage: ", row["message"])
            #return row["message"]
            msg_less_stem_equal_lemma = row["message"] 
            found_msg_1_flag = True
        
        if len(lemmatized) < len(original_lemmatized) and len(stemmed) == len(original_stemmed):
             print(f"Message at index {idx} causes a change in lemmatized tokens but not in stemmed tokens. \nmessage:", row["message"])
            # return row["message"]  # Return the message that causes the change
             msg_less_lemma_equal_stem = row["message"] 
             found_msg_2_flag = True

        if (found_msg_1_flag == True and found_msg_2_flag== True):
            break # Break the loop if such messages found

            

    if  (found_msg_1_flag == False and found_msg_2_flag == False):      
        print("No message found that meets the criteria.")
    
  
   
    #return None


find_message_to_remove(df_whatsapp_processed_text)



No message found that meets the criteria.


#### Results 17.8-9:

For Hebrew messages, due to the limitations of the NLTK and library in handling the Hebrew language effectively, the results of lemmatization and stemming are often identical. This overlap makes it impossible to differentiate between the number of lemmas and stems within the same segment of text.


#### 17.10-11

Both NLTK and spaCy lack support for lemmatization and stemming in the Hebrew language. As a result, their functionality for Hebrew text is limited to tokenization, producing identical outputs for tokenization, lemmatization, and stemming operations.